source/_posts/2018-02-26-URLComponents.md

```diff
@@ -0,0 +1,367 @@
+
+    return urlComponents.url
+}
+```
+
+First we use the `urlString` to create a new `URLComponents` object. (Note that we create the `urlComponents` as a `var` as we'll need to write to it later on.)
+
+Next we create an array of `URLQueryItem` objects and append our only required parameter, the `q` parameter.
+
+Then we loop through our `optionalParameters` dictionary, map the elements to `URLQueryItem` objects and append them to our existing array.
+
+Finally, we ask `URLComponents` to try and return a `URL` for us.
+
+Simple and beautiful, all the hard work has been delegated onwards to `URLComponents`.
+
+Lets see how it works:
+
+```swift
+if let url = searchURL(with: "Sopranos") {
+    print(url) //gives us: https://showsknownfrom.tv/search?q=Sopranos
+}
+
+if let url = searchURL(with: "Sopranos", optionalParameters: ["order" : "desc", "number" : "100"]) {
+    print(url) //gives us: https://showsknownfrom.tv/search?q=Sopranos&number=100&order=desc
+}
+
+if let url = searchURL(with: "Halt and Catch Fire", optionalParameters: ["order" : "desc", "number" : "100"]) {
+    print(url) //gives us: https://showsknownfrom.tv/search?q=Halt%20and%20Catch%20Fire&number=100&order=desc
+} else {
+    print("no URL")
+}
+```
+
+Perfect!
+
+### Decomposing URLs
+Remember our example from before?
+
+Now that we know about `URLQueryItem`s, checking if a query parameter exists is soooo much easier:
+
+```swift
+func receivedUrl(_ url: URL, contains parameter: String) -> String? {
+    guard
+        let urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: true),
+        let queryItems = urlComponents.queryItems
+    else {
+        return nil
+    }
+
+    let items = queryItems.filter({ $0.name == parameter })
+    return items.first?.value
+}
+
+if let url = URL(string: "https://showsknownfrom.tv/shows/12345678?featured=true") {
+  if let value = receivedUrl(url, contains: "featured") {
+    print("found: \(value)")
+  }
+}
+```
+
+Pretty right? At least compared to the initial version.
+
+## Composing and Decomposing URLs - The Easier Way
+Well, we made it this far! But can we do better? Of course we can!
+
+Lets write an `extension` to `URL` so we can append query parameters and also ask if a URL contains a query parameter:
+
+```swift
+extension URL {
+  func append(queryParameters: [String: String]) -> URL? {
+      guard var urlComponents = URLComponents(url: self, resolvingAgainstBaseURL: true) else {
+          return nil
+      }
+
+      let urlQueryItems = queryParameters.map{
+          return URLQueryItem(name: $0, value: $1)
+      }
+      urlComponents.queryItems = urlQueryItems
+      return urlComponents.url
+  }
+
+  func value(forParameter name: String) -> String? {
+    guard let urlComponents = URLComponents(url: self, resolvingAgainstBaseURL: true),
+        let queryItems = urlComponents.queryItems else {
+            return nil
+    }
+    let items = queryItems.filter({ $0.name == name })
```

Comment:
you can leave out the '(' and ')' or even write something like:
```
return URLComponents(url: self, resolvingAgainstBaseURL: true)?
    .queryItems?
    .filter { $0.name == name }?
    .first?
    .value
```

But that's a lot for one statement :P